lib/utils/index.ts

```diff
@@ -15,17 +15,88 @@ export const DEFAULT_PGSTAC_VERSION = "0.9.5";
 export function resolveLambdaCode(
   userCode?: lambda.Code,
   dockerBuildPath?: string,
-  dockerBuildOptions?: lambda.DockerBuildAssetOptions
+  dockerBuildOptions?: lambda.DockerBuildAssetOptions,
 ): lambda.Code {
   if (userCode) {
     return userCode;
   }
 
   if (!dockerBuildPath || !dockerBuildOptions) {
     throw new Error(
-      "dockerBuildPath and dockerBuildOptions are required when no custom code is provided"
+      "dockerBuildPath and dockerBuildOptions are required when no custom code is provided",
     );
   }
 
   return lambda.Code.fromDockerBuild(dockerBuildPath, dockerBuildOptions);
 }
+
+/**
+ * Type guard to check if a value has the secretBootstrapper property.
+ *
+ * This is used to detect PgStacDatabase constructs with PgBouncer enabled,
+ * which expose a secretBootstrapper CustomResource that completes after
+ * the database secret is updated with PgBouncer connection information.
+ *
+ * @param value - Value to check
+ * @returns true if value has a defined secretBootstrapper property
+ */
+function hasSecretBootstrapper(
+  value: any,
+): value is { secretBootstrapper: CustomResource } {
+  return (
+    value &&
+    typeof value === "object" &&
+    "secretBootstrapper" in value &&
+    value.secretBootstrapper !== undefined
+  );
+}
+
+/**
+ * Extracts database dependencies from a database construct if available.
+ *
+ * For PgStacDatabase with PgBouncer enabled, returns the secretBootstrapper
+ * CustomResource which ensures the database secret is fully initialized before
+ * dependent resources are created.
+ *
+ * This is critical for SnapStart Lambda functions, as the snapshot must not
+ * be created until the secret contains the correct connection information.
+ *
+ * @param db - Database construct (may be PgStacDatabase, IDatabaseInstance, etc.)
+ * @returns Array of CustomResources to depend on (empty if none available)
+ */
+export function extractDatabaseDependencies(db: any): CustomResource[] {
+  if (hasSecretBootstrapper(db)) {
+    return [db.secretBootstrapper];
+  }
+  return [];
```

Comment:
Just checking, so if SnapStart isn't set then we don't necessarily need db.secretBootstrapper?